src/quadexpr.jl

```diff
@@ -61,25 +60,59 @@ function Base.isequal{T,S}(q::GenericQuadExpr{T,S},other::GenericQuadExpr{T,S})
     return true
 end
 
+# Check if two QuadExprs are equal regardless of the order, and after merging duplicates
+# Mostly useful for testing.
+function isequal_canonical(quad::GenericQuadExpr{CoefType,VarType}, other::GenericQuadExpr{CoefType,VarType}) where {CoefType,VarType}
+    function canonicalize(q)
+        d = Dict{Set{VarType},CoefType}()
+        @assert length(q.qvars1) == length(q.qvars2) == length(q.qcoeffs)
+        for (v1,v2,c) in zip(q.qvars1, q.qvars2, q.qcoeffs)
+            vset = Set((v1,v2))
+            d[vset] = c + get(d, vset, zero(CoefType))
+        end
+        return d
+    end
+    d1 = canonicalize(quad)
+    d2 = canonicalize(other)
+    return isequal(d1,d1) && isequal_canonical(quad.aff, other.aff)
+end
+
 # Alias for (Float64, Variable)
 const QuadExpr = GenericQuadExpr{Float64,Variable}
 Base.convert(::Type{QuadExpr}, v::Union{Real,Variable,AffExpr}) = QuadExpr(Variable[], Variable[], Float64[], AffExpr(v))
 QuadExpr() = zero(QuadExpr)
 
-# function setobjective(m::Model, sense::Symbol, q::QuadExpr)
-#     m.obj = q
-#     if m.internalModelLoaded
-#         if method_exists(MathProgBase.setquadobjterms!, (typeof(m.internalModel), Vector{Cint}, Vector{Cint}, Vector{Float64}))
-#             verify_ownership(m, m.obj.qvars1)
-#             verify_ownership(m, m.obj.qvars2)
-#             MathProgBase.setquadobjterms!(m.internalModel, Cint[v.col for v in m.obj.qvars1], Cint[v.col for v in m.obj.qvars2], m.obj.qcoeffs)
-#         else
-#             isa(m.internalModel, MathProgBase.AbstractLinearQuadraticModel) && Base.warn_once("Solver does not support adding a quadratic objective to an existing model. JuMP's internal model will be discarded.")
-#             m.internalModelLoaded = false
-#         end
-#     end
-#     setobjectivesense(m, sense)
-# end
+function MOI.ScalarQuadraticFunction(q::QuadExpr)
+    return MOI.ScalarQuadraticFunction(instancereference.(q.aff.vars), q.aff.coeffs, instancereference.(q.qvars1), instancereference.(q.qvars2), q.qcoeffs, q.aff.constant)
+end
+
+function QuadExpr(m::Model, f::MOI.ScalarQuadraticFunction)
+    return QuadExpr(Variable.(m,f.quadratic_rowvariables), Variable.(m, f.quadratic_colvariables), f.quadratic_coefficients, AffExpr(Variable.(m,f.affine_variables), f.affine_coefficients, f.constant))
+end
+
+function setobjective(m::Model, sense::Symbol, a::QuadExpr)
+    @assert !m.solverinstanceattached # TODO
+    if sense == :Min
```

Comment:
Why don't we do
```julia
function setobjective(m::Model, sense::Symbol, a::AbstractJuMPScalar)
    ...
    MOI.set!(m.instance, MOI.ObjectiveFunction(), JuMPScalar2MOIfunction(a))
    nothing
end
```
Of course the name `JuMPScalar2MOIfunction` should be replaced by something better.

Reply:
Makes sense, but I'll leave it for a future refactor.